include/exadg/utilities/optional_function.h

```diff
@@ -0,0 +1,54 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2024 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef EXADG_AERO_UTILITIES_OPTIONAL_FUNCTION_H_
+#define EXADG_AERO_UTILITIES_OPTIONAL_FUNCTION_H_
+
+#include <deal.II/base/function.h>
+
+namespace ExaDG
+{
+namespace Utilities
+{
+/**
+ * This is a @c dealii::Function, that provides information if it has
+ * to be evaluated at a given time.
+ */
+
+template<int dim, typename Number = double>
+class OptionalFunction : public dealii::Function<dim, Number>
+{
+public:
+  using time_type = typename dealii::Function<dim, Number>::time_type;
+
+  OptionalFunction(const unsigned int n_components = 1, const time_type initial_time = 0.0)
+    : dealii::Function<dim, Number>(n_components, initial_time)
+  {
+  }
+
+  virtual bool
+  needs_evaluation_at_time(double const new_time) const = 0;
```

Comment:
In case of "fade out in space", this function needs to return `true` throughout the simulation, right? The more I think about it, I am less sure whether the design is good currently.

Overall, I do not understand the trade-offs made regarding computational efficiency, by which the current design seems to be influenced. You introduced `needs_evaluation_at_time()` in order to evaluate the function not too often if I understood it correctly. You also write above that "fade out in space" is used in rare cases. This would mean that in the majority of cases, the blending function is called in very quadrature point, even though a global call (once for the whole mesh) would be enough? Performance-wise, I consider calling the function unnecessarily for every q-point more relevant than calling it un-necessarily once every time step. Against this background, it would argue we need a function `bool varies_in_space()` (which is the conditional on callling `Function::value()`) and a function `double get_time_factor()` (returing the temporal factor by which to multiply the source term) for optimal performance (but no function `needs_evaluation_at_time()` combined with `Function::value()` always evaluating spatial and temporal dimensions). The alternative is option 3 described above (one single `Function::value()` for all spatial and temporal dimensions; note that this function will only be called if `data.blend_in = true`, so the cases without blending in space or time will not be slowed down by this additional functionality).

Reply:
> In case of "fade out in space", this function needs to return true throughout the simulation, right? 

Yes

I see what you mean. What do you think about completely seperating the spatial and temporal blending? We could have a function that deals with the spatial rampup and one that deals with the spatial blending. It is not intuitive to have a field function that returns a constant time factor for me. The question would however be, where the function for the temporal blending should live?

Reply:
Basically, a `dealii::Function` treats both spatial and temporal components. The interface of `Function::value()` does not look like this, and I can only guess where this comes from. Maybe one does not want to see a time argument for steady-state problems, In any case, this has been a design decision on the dealii side we have to live with. Nevertheless, this implies that we basically (can/should) use the same function to treat both spatial and temporal variations, unless we consider this too inefficient because in the majority of cases this function would be called unnecessarily. However, I believe this will also restrict functionality.

I think separating the spatial and temporal blending might be two restrictive. I think about two source terms in two different parts of the domain that become active at different times. How to realise this? The final source term is not simply `time_factor * spatial_factor`, while a normal `dealii::Function::value()` in combination with `dealii::Function::set_time()` would have this capability naturally.

Reply:
I had a hard time to separate the comments above. I tried to resolve every issue in [refracture application of scaling factor](https://github.com/exadg/exadg/pull/644/commits/1026d662abd8a15b3a60b22b307db977ea36f10b)